File: ai/fsm.lua
function createFSM()
    local patrol = PatrolState(waypoints);
    patrol.addTransition(Transition.SawPlayer, FSMStateID.Chasing)
    patrol.addTransition(Transition.NoHealth, FSMStateID.Dead)

    local dead = DeadState()
    dead.addTransition(Transition.NoHealth, FSMStateID.Dead)

    addFSMState(patrol)
end

File: ai/fsm.py
#!/usr/bin/env python
# encoding: utf-8


transitions = ['saw_player', 'reach_player', 'lost_player', 'no_health']
states = ['patrolling', 'chasing', 'attacking', 'dead']

_x = lambda x: dict((k, v) for v, k in enumerate(x))

TRANSITION = _x(transitions)
FSMStateID = _x(states)


class FSMState(object):
    def reason():
        pass

    def act():
        pass


def PatrolState(FSMState):
    def reason(player, npc):
        if npc.position >= player.position:
            npc.set_transition(TRANSITION['saw_player'])

    def act(player, npc):
        print('go go go patroll')
